Support `clobber_abi` in AVR inline assembly

jfrimmel opened the pull request:
This PR implements the `clobber_abi` part necessary to eventually stabilize the inline assembly for AVR. This is tracked in #93335.
This is heavily inspired by the sibling-PR #131310 for the MSP430. I've explained my reasoning in the first commit message in detail, which is reproduced below for easier reviewing:

This follows the [ABI documentation] of AVR-GCC:

> The [...] call-clobbered general purpose registers (GPRs) are registers that might be destroyed (clobbered) by a function call.
>
> - **R18–R27, R30, R31**
>
>   These GPRs are call clobbered. An ordinary function may use them without restoring the contents. [...]
>
> - **R0, T-Flag**
>
>   The temporary register and the T-flag in SREG are also call-clobbered, but this knowledge is not exposed explicitly to the compiler (R0 is a fixed register).

Therefore this commit lists the aforementioned registers `r18–r27`, `r30` and `r31` as clobbered registers. Since the `r0` register (listed above as well) is not available in inline assembly at all (potentially because the AVR-GCC considers it a fixed register causing the register to never be used in register allocation and LLVM adopting this), there is no need to list it in the clobber list (the `r0`-variant is not even available). A comment was added to ensure, that the `r0` gets added to the clobber-list once the register gets usable in inline ASM.
Since the SREG is normally considered clobbered anyways (unless the user supplies the `preserve_flags`-option), there is no need to explicitly list a bit in this register (which is not possible to list anyways).

Note, that this commit completely ignores the case of interrupts (that are described in the ABI-specification), since every register touched in an ISR need to be saved anyways.

[ABI documentation]: https://gcc.gnu.org/wiki/avr-gcc#Call-Used_Registers



r? @Amanieu

@rustbot label +O-AVR

jfrimmel committed:
Support `clobber_abi` for AVR inline assembly

This commit adds the relevant registers to the list of clobbered regis-
ters (part of #93335). This follows the [ABI documentation] of AVR-GCC:

> The [...] call-clobbered general purpose registers (GPRs) are
> registers that might be destroyed (clobbered) by a function call.
>
> - **R18–R27, R30, R31**
>
>   These GPRs are call clobbered. An ordinary function may use them
>   without restoring the contents. [...]
>
> - **R0, T-Flag**
>
>   The temporary register and the T-flag in SREG are also call-
>   clobbered, but this knowledge is not exposed explicitly to the
>   compiler (R0 is a fixed register).

Therefore this commit lists the aforementioned registers `r18–r27`,
`r30` and `r31` as clobbered registers. Since the `r0` register (listed
above as well) is not available in inline assembly at all (potentially
because the AVR-GCC considers it a fixed register causing the register
to never be used in register allocation and LLVM adopting this), there
is no need to list it in the clobber list (the `r0`-variant is not even
available). A comment was added to ensure, that the `r0` gets added to
the clobber-list once the register gets usable in inline ASM.
Since the SREG is normally considered clobbered anyways (unless the user
supplies the `preserve_flags`-option), there is no need to explicitly
list a bit in this register (which is not possible to list anyways).

Note, that this commit completely ignores the case of interrupts (that
are described in the ABI-specification), since every register touched in
an ISR need to be saved anyways.

[ABI documentation]: https://gcc.gnu.org/wiki/avr-gcc#Call-Used_Registers

jfrimmel committed:
Add test case for the clobber options

Amanieu commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -106,6 +106,7 @@ def_regs! {
             "the stack pointer cannot be used as an operand for inline asm",
         #error = ["r0", "r1", "r1r0"] =>
             "r0 and r1 are not available due to an issue in LLVM",
+            // if this issue is resolved, then R0 has to be added to clobbered registers
```
We need to carefully keep track of LLVM here: if LLVM ceases to treat these registers as reserved then we must immediately add them to the set of clobbered registers, otherwise incorrect code will be generated. This comment should be updated to reflect this.

Also, while reviewing the GCC docs, it seems that R1 always holds a value of 0. If this is a guarantee that asm can use then it should be documented, similar to how on x86 we guarantee the direction flag is always clear when entering or exiting inline asm.

Amanieu replied:
It's enough to just reword the comment.

jfrimmel replied:
I've reworded the comment, please have a look if this works for you.

Regarding the R1 register content: the unstable-book already documents, that the value needs to be restored if altered:
https://github.com/rust-lang/rust/blob/17a19e684cdf3ca088af8b4da6a6209d128913f4/src/doc/unstable-book/src/language-features/asm-experimental-arch.md?plain=1#L156

jfrimmel commented:
@rustbot labels: +A-inline-assembly
@rustbot ready

taiki-e commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -106,6 +106,11 @@ def_regs! {
             "the stack pointer cannot be used as an operand for inline asm",
         #error = ["r0", "r1", "r1r0"] =>
             "r0 and r1 are not available due to an issue in LLVM",
```
(IIUC since r1 is a zero register, r1 and r1r0 cannot be used as an operand for inline asm, regardless of an LLVM issue.)

jfrimmel replied:
I think, the fact that r1 is the zero-reg _is_ the cause for the registers not being available for inline asm (although I'm not sure). For now, I think this is fine (please correct me if you think otherwise!).

taiki-e replied:
> For now, I think this is fine (please correct me if you think otherwise!).

Agreed. We may need to figure it out in some way before stabilization, but at least it should not be a blocker of clobber_abi implementation.

Amanieu replied:
It would actually be nice to make the error message here clearer than "due to an issue in LLVM". The reason is that r0 is reserved as a scratch register and r1 is reserved as a constant zero register by LLVM.

taiki-e commented on tests/codegen/asm-clobber_abi-avr.rs:
```diff
@@ -0,0 +1,43 @@
+//@ assembly-output: emit-asm
```
Could you move this file into tests/codegen/asm directory like https://github.com/rust-lang/rust/pull/132456?

jfrimmel replied:
Done :+1:

jfrimmel committed:
Move & rename test case to match naming of #132456

taiki-e commented on tests/codegen/asm/avr-clobbes.rs:
```diff
@@ -0,0 +1,43 @@
+//@ assembly-output: emit-asm
```
> avr-clobbes.rs

"r" appears to have been accidentally dropped from the filename. (avr-clobbers.rs should be correct)

jfrimmel replied:
Well spotted, sorry! Fixed in the force-push

taiki-e commented:
Btw, ATTiny (TinyAVR) appears to have a different rules (r0-r15 are unavailable, r16-r17 are similar to r0-r1 in AVR, r18-r19 and r28-r29 are call-saved).
- https://gcc.gnu.org/wiki/avr-gcc#Reduced_Tiny
- https://github.com/llvm/llvm-project/blob/fb7bf7a5acc65be44fc546f282942b91472553b3/llvm/lib/Target/AVR/AVRRegisterInfo.cpp#L70

jfrimmel commented:
Hm, I think in order to properly address the reduced tiny cores, one would need to wait for #131651 and parse the `-Ctarget-cpu`-flag to detect such a core. I suppose, this change would also be necessary for the general inline assembly and thus is somewhat out of scope of this PR. I see two options:
1. merge this as-is and perform the necessary changes in a follow-up PR
2. close this and wait for #131651 to be merged, then fix the reduced tiny cores and then the clobber ABI.

taiki-e commented:
> I suppose, this change would also be necessary for the general inline assembly and thus is somewhat out of scope of this PR.

I think it is indeed true that inline assembly support for AVRTiny is already broken (registers that should be reserved are not reserved) and therefore does not need to be fixed in this PR.

> parse the `-Ctarget-cpu`-flag to detect such a core.

I think it is better to refer to the [`tinyencoding` target feature](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0/llvm/lib/Target/AVR/AVRDevices.td#L119) as [LLVM does](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0/llvm/lib/Target/AVR/AVRRegisterInfo.cpp#L71), instead of parsing the target-cpu.
(This could be implemented in a similar way to https://github.com/rust-lang/rust/pull/132472#discussion_r1826554303.)

Amanieu commented:
Could you please fix the wording as suggested in https://github.com/rust-lang/rust/pull/131323#discussion_r1831790068, then this can be merged.

Amanieu commented:
@rustbot author

jfrimmel committed:
Reword error message of reserved AVR registers

Those are reserved as per the GCC (and thus LLVM) ABI, which is distinct from an
issue. The rewording was requested in this [review].

[review]: https://github.com/rust-lang/rust/pull/131323#issuecomment-2479178721

jfrimmel commented:
Sorry, I've missed your review feedback. I've reworded the line in question.

@rustbot ready

bors commented:
:pushpin: Commit 297f2a5acc5fa841ae5f6a10749448875acae478 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 297f2a5acc5fa841ae5f6a10749448875acae478 -->
<!-- homu: {"type":"Approved","sha":"297f2a5acc5fa841ae5f6a10749448875acae478","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #133568) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jfrimmel commented:
rebased

@rustbot ready

Amanieu commented:
@bors r+
